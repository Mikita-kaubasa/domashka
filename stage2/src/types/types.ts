/**
 * @fileoverview TypeScript type definitions for the user management application.
 * This module contains all the core interfaces and types used throughout the application,
 * including user data structures, address information, and company details.
 * 
 * @module types
 */

/**
 * Geographic coordinates for a location.
 * 
 * @interface Geo
 * @property {string} lat - Latitude coordinate as a string
 * @property {string} lng - Longitude coordinate as a string
 * 
 * @example
 * ```typescript
 * const coordinates: Geo = {
 *   lat: "40.7128",
 *   lng: "-74.0060"
 * };
 * ```
 */
export interface Geo {
  lat: string;
  lng: string;
}

/**
 * Complete address information for a user.
 * 
 * @interface Address
 * @property {string} street - Street name and number
 * @property {string} suite - Suite, apartment, or unit number
 * @property {string} city - City name
 * @property {string} zipcode - Postal/ZIP code
 * @property {Geo} geo - Geographic coordinates of the address
 * 
 * @example
 * ```typescript
 * const address: Address = {
 *   street: "123 Main St",
 *   suite: "Apt 4B",
 *   city: "New York",
 *   zipcode: "10001",
 *   geo: { lat: "40.7128", lng: "-74.0060" }
 * };
 * ```
 */
export interface Address {
  street: string;
  suite: string;
  city: string;
  zipcode: string;
  geo: Geo;
}

/**
 * Company information associated with a user.
 * 
 * @interface Company
 * @property {string} name - Company name
 * @property {string} catchPhrase - Company slogan or catchphrase
 * @property {string} bs - Business strategy or description
 * 
 * @example
 * ```typescript
 * const company: Company = {
 *   name: "Acme Corp",
 *   catchPhrase: "Making the world a better place",
 *   bs: "harness real-time e-markets"
 * };
 * ```
 */
export interface Company {
  name: string;
  catchPhrase: string;
  bs: string;
}

/**
 * Complete user information structure.
 * 
 * This is the main data structure representing a user in the application.
 * It contains all personal, contact, and professional information.
 * 
 * @interface User
 * @property {number} id - Unique identifier for the user
 * @property {string} name - Full name of the user
 * @property {string} username - Unique username for the user
 * @property {string} email - Email address of the user
 * @property {Address} address - Complete address information
 * @property {string} phone - Phone number (can be in various formats)
 * @property {string} website - Personal or company website URL
 * @property {Company} company - Associated company information
 * 
 * @example
 * ```typescript
 * const user: User = {
 *   id: 1,
 *   name: "John Doe",
 *   username: "johndoe",
 *   email: "john@example.com",
 *   address: {
 *     street: "123 Main St",
 *     suite: "Apt 4B",
 *     city: "New York",
 *     zipcode: "10001",
 *     geo: { lat: "40.7128", lng: "-74.0060" }
 *   },
 *   phone: "1-770-736-8031 x56442",
 *   website: "johndoe.org",
 *   company: {
 *     name: "Acme Corp",
 *     catchPhrase: "Making the world a better place",
 *     bs: "harness real-time e-markets"
 *   }
 * };
 * ```
 */
export interface User {
  id: number;
  name: string;
  username: string;
  email: string;
  address: Address;
  phone: string;
  website: string;
  company: Company;
} 