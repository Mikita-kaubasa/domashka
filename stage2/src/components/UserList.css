.user-list-container {
    max-width: 1100px;
    margin: 0 auto;
    padding: 2rem 0 0 0;
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
        'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
        sans-serif;
}

.user-list-title {
    color: #2c3e50;
    font-size: 2rem;
    font-weight: 600;
    margin-bottom: 2rem;
    text-align: center;
}

.user-table-wrapper {
    background: white;
    border-radius: 12px;
    box-shadow: 0 2px 8px 0 rgba(0, 0, 0, 0.04);
    overflow-x: visible;
    width: 100%;
    min-width: 0;
}

.user-table.modern {
    width: 100%;
    min-width: 0;
    border-collapse: separate;
    border-spacing: 0;
    background: white;
    font-size: 1rem;
    table-layout: auto;
}

.table-header {
    background: #f7f8fa;
    color: #6b7280;
    font-weight: 600;
    font-size: 0.95rem;
    letter-spacing: 0.04em;
    text-transform: uppercase;
}

.table-header th {
    padding: 1.1rem 0.8rem;
    text-align: left;
    font-weight: 600;
    border: none;
    white-space: nowrap;
}

.table-row {
    transition: background 0.15s;
    cursor: pointer;
    border: none;
}

.table-row:hover {
    background: #f3f4f6;
}

.table-row td {
    padding: 1.1rem 0.8rem;
    vertical-align: middle;
    border: none;
    font-size: 1rem;
    white-space: normal;
    text-overflow: initial;
    overflow: visible;
    max-width: none;
}

.user-info {
    min-width: 0;
    max-width: none;
    display: flex;
    flex-direction: column;
    gap: 0.1rem;
    word-break: break-word;
}

.user-name {
    font-weight: 700;
    color: #222;
    font-size: 1.08rem;
    line-height: 1.2;
    margin-bottom: 0.1rem;
    white-space: normal;
    overflow: visible;
    text-overflow: initial;
}

.user-email {
    color: #6b7280;
    font-size: 0.97rem;
    font-weight: 400;
    white-space: normal;
    overflow: visible;
    text-overflow: initial;
}

.user-address,
.user-company,
.user-phone,
.user-website {
    color: #222;
    font-size: 0.97rem;
    font-weight: 400;
    white-space: normal;
    overflow: visible;
    text-overflow: initial;
    max-width: none;
    word-break: break-word;
}

.user-phone a,
.user-website a {
    color: #2563eb;
    text-decoration: none;
    font-weight: 500;
    transition: color 0.2s;
}

.user-phone a:hover,
.user-website a:hover {
    color: #1d4ed8;
    text-decoration: underline;
}

.user-action {
    text-align: center;
    min-width: 36px;
    width: 44px;
    max-width: 60px;
    padding: 0 !important;
    box-sizing: border-box;
}

.action-btn {
    background: none;
    border: none;
    box-shadow: none;
    outline: none;
    cursor: pointer;
    min-width: 32px;
    min-height: 32px;
    padding: 0;
    margin: 0 auto;
    display: flex;
    align-items: center;
    justify-content: center;
}

.action-x {
    color: #ef4444;
    font-size: 1.35rem;
    font-weight: 700;
    line-height: 1;
    transition: color 0.15s;
    user-select: none;
}

.action-btn:hover .action-x,
.action-btn:focus .action-x {
    color: #b91c1c;
}

/* Responsive design */
@media (max-width: 1024px) {
    .user-list-container {
        padding: 1rem;
    }

    .user-table.modern {
        font-size: 0.97rem;
    }

    .table-header th,
    .table-row td {
        padding: 0.55rem 0.3rem;
    }

    .user-info,
    .user-address,
    .user-company {
        min-width: 0;
        max-width: none;
    }

    .user-action {
        min-width: 36px;
        width: 40px;
        max-width: 50px;
    }
}

@media (max-width: 900px) {
    .user-table.modern {
        font-size: 0.93rem;
    }

    .table-header th,
    .table-row td {
        padding: 0.45rem 0.2rem;
    }

    .user-info {
        min-width: 0;
        max-width: none;
    }

    .user-action {
        min-width: 36px;
        width: 36px;
        max-width: 44px;
    }
}

@media (max-width: 700px) and (min-width: 601px) {
    .user-table.modern tr.table-row {
        flex-wrap: wrap;
    }

    .user-info {
        min-width: 0;
        max-width: none;
    }

    .user-action {
        min-width: 36px;
        width: 36px;
        max-width: 44px;
        display: flex;
        align-items: flex-start;
        justify-content: flex-end;
        padding-top: 0.2rem !important;
    }

    .action-btn {
        min-width: 32px;
        min-height: 32px;
    }
}

@media (max-width: 600px) {
    .user-list-container {
        padding: 0.5rem 0 0 0;
    }

    .user-table.modern {
        display: block;
        border: none;
        font-size: 1rem;
    }

    .user-table.modern thead {
        display: none;
    }

    .user-table.modern tbody {
        display: block;
    }

    .user-table.modern tr.table-row {
        display: flex;
        flex-direction: column;
        background: #fff;
        border-radius: 12px;
        box-shadow: 0 1px 4px 0 rgba(0, 0, 0, 0.06);
        margin-bottom: 1.1rem;
        padding: 1.1rem 1rem 1rem 1rem;
        position: relative;
        min-width: 0;
        max-width: 100vw;
    }

    .user-table.modern td {
        display: block;
        padding: 0.3rem 0 0.3rem 0;
        font-size: 1rem;
        max-width: 100vw;
        white-space: normal;
        border: none;
        background: none;
        box-shadow: none;
        overflow: visible;
    }

    .user-info {
        min-width: 0;
        max-width: 100vw;
        margin-bottom: 0.2rem;
    }

    .user-name {
        font-size: 1.13rem;
        margin-bottom: 0.1rem;
    }

    .user-email {
        font-size: 1rem;
        margin-bottom: 0.5rem;
    }

    .user-address,
    .user-company,
    .user-phone,
    .user-website {
        font-size: 1rem;
        max-width: 100vw;
        margin-bottom: 0.3rem;
        white-space: normal;
    }

    .user-action {
        position: absolute;
        top: 1.1rem;
        right: 1.1rem;
        width: auto;
        min-width: 0;
        padding: 0;
        text-align: right;
        z-index: 2;
    }

    .action-btn {
        padding: 0.2rem 0.5rem;
        font-size: 1.5rem;
        background: none;
        border-radius: 50%;
        border: none;
        box-shadow: none;
        outline: none;
        transition: background 0.15s;
    }

    .action-btn:active {
        background: #f3f4f6;
    }

    .action-x {
        font-size: 1.5rem;
    }
}

@media (max-width: 1024px) {
    .user-list-container {
        padding: 1rem;
    }

    .table-header th,
    .table-row td {
        padding: 0.75rem 0.5rem;
    }

    .user-info,
    .user-address,
    .user-company {
        min-width: auto;
    }
}

@media (max-width: 768px) {
    .user-table-wrapper {
        border-radius: 8px;
    }

    .user-list-title {
        font-size: 1.5rem;
        margin-bottom: 1.5rem;
    }

    .table-header th {
        font-size: 0.8rem;
        padding: 0.75rem 0.5rem;
    }

    .table-row td {
        padding: 0.75rem 0.5rem;
    }

    .user-name,
    .company-name {
        font-size: 0.9rem;
    }

    .user-email,
    .user-address,
    .company-catchphrase {
        font-size: 0.8rem;
    }
}

@media (max-width: 640px) {
    .user-table {
        font-size: 0.8rem;
    }

    .table-header th {
        font-size: 0.75rem;
        padding: 0.5rem 0.25rem;
    }

    .table-row td {
        padding: 0.5rem 0.25rem;
    }
}

.user-toast {
    position: fixed;
    left: 50%;
    bottom: 2.5rem;
    transform: translateX(-50%);
    background: #222;
    color: #fff;
    padding: 0.75rem 1.5rem;
    border-radius: 8px;
    font-size: 1.05rem;
    font-weight: 500;
    box-shadow: 0 4px 16px 0 rgba(0, 0, 0, 0.12);
    z-index: 2000;
    opacity: 0.97;
    animation: fadeInToast 0.2s;
}

@keyframes fadeInToast {
    from {
        opacity: 0;
        transform: translateX(-50%) translateY(20px);
    }

    to {
        opacity: 0.97;
        transform: translateX(-50%) translateY(0);
    }
}

.confirm-backdrop {
    background: rgba(0, 0, 0, 0.18);
    z-index: 3000;
}

.confirm-modal {
    background: #fff;
    border-radius: 14px;
    box-shadow: 0 8px 32px 0 rgba(0, 0, 0, 0.18);
    padding: 2rem 2.5rem 1.5rem 2.5rem;
    min-width: 320px;
    max-width: 90vw;
    margin: 0 auto;
    display: flex;
    flex-direction: column;
    align-items: center;
    gap: 1.2rem;
    animation: fadeIn 0.2s;
}

.confirm-title {
    font-size: 1.25rem;
    font-weight: 700;
    color: #222;
    margin-bottom: 0.2rem;
}

.confirm-message {
    color: #444;
    font-size: 1.05rem;
    margin-bottom: 0.5rem;
    text-align: center;
}

.confirm-actions {
    display: flex;
    gap: 1.2rem;
    margin-top: 0.5rem;
}

.confirm-btn {
    font-size: 1rem;
    font-weight: 500;
    border: none;
    border-radius: 7px;
    padding: 0.6rem 1.4rem;
    cursor: pointer;
    transition: background 0.15s, color 0.15s;
}

.confirm-btn.cancel {
    background: #f3f4f6;
    color: #222;
}

.confirm-btn.cancel:hover,
.confirm-btn.cancel:focus {
    background: #e5e7eb;
}

.confirm-btn.delete {
    background: #ef4444;
    color: #fff;
}

.confirm-btn.delete:hover,
.confirm-btn.delete:focus {
    background: #b91c1c;
}