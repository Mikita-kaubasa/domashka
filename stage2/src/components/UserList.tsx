/**
 * @fileoverview User list component that displays users in a table format with interactive features.
 * This component provides a comprehensive user management interface with the ability to view,
 * delete users, and open detailed user information in modals.
 * 
 * @module UserList
 * @requires React
 * @requires ../types/types
 * @requires ../utils/formatters
 * @requires ./ConfirmDeleteModal
 * @requires ./UserDetailsModal
 * @requires ./UserList.css
 */

import React, { Suspense, lazy, memo, useCallback, useMemo, useState } from 'react';
import { User } from '../types/types';
import { formatAddress, formatWebsite } from '../utils/formatters';
import ConfirmDeleteModal from './ConfirmDeleteModal';
import './UserList.css';

// Lazy load the modal component for better performance
const UserDetailsModal = lazy(() => import('./UserDetailsModal'));

/**
 * Props for the UserList component.
 * 
 * @interface UserListProps
 * @property {User[]} users - Array of user objects to display
 */
interface UserListProps {
    users: User[];
}

/**
 * User list component with interactive features.
 * 
 * This component provides:
 * - Tabular display of user information
 * - Click-to-view user details in a modal
 * - Delete user functionality with confirmation
 * - Responsive design with modern styling
 * - Accessibility features (ARIA labels, keyboard navigation)
 * - Lazy loading of modal components
 * 
 * @component
 * @param {UserListProps} props - Component props
 * @returns {JSX.Element} User list interface
 * 
 * @example
 * ```tsx
 * import UserList from './components/UserList';
 * 
 * function App() {
 *   const users = [/* user data *\/];
 *   return <UserList users={users} />;
 * }
 * ```
 */
const UserList: React.FC<UserListProps> = memo(({ users: initialUsers }) => {
    const [users, setUsers] = useState<User[]>(initialUsers);
    const [selectedUser, setSelectedUser] = useState<User | null>(null);
    const [isModalOpen, setIsModalOpen] = useState(false);
    const [deletedMsg, setDeletedMsg] = useState<string | null>(null);
    const [confirmOpen, setConfirmOpen] = useState(false);
    const [userToDelete, setUserToDelete] = useState<User | null>(null);

    /**
     * Handles row click to open user details modal.
     * 
     * @param {User} user - The user object to display in the modal
     * 
     * @example
     * ```typescript
     * handleRowClick(user);
     * ```
     */
    const handleRowClick = useCallback((user: User) => {
        setSelectedUser(user);
        setIsModalOpen(true);
    }, []);

    /**
     * Closes the user details modal and resets selected user.
     * 
     * @example
     * ```typescript
     * handleCloseModal();
     * ```
     */
    const handleCloseModal = useCallback(() => {
        setIsModalOpen(false);
        setSelectedUser(null);
    }, []);

    /**
     * Initiates the delete user process by opening confirmation modal.
     * 
     * @param {User} user - The user to be deleted
     * @param {React.MouseEvent} e - Optional mouse event (used to prevent row click)
     * 
     * @example
     * ```typescript
     * handleDeleteUser(user, event);
     * ```
     */
    const handleDeleteUser = useCallback((user: User, e?: React.MouseEvent) => {
        if (e) e.stopPropagation();
        setUserToDelete(user);
        setConfirmOpen(true);
    }, []);

    /**
     * Confirms user deletion and removes from the list.
     * 
     * This function:
     * - Removes the user from the local state
     * - Shows a temporary success message
     * - Closes the confirmation modal
     * 
     * @example
     * ```typescript
     * handleConfirmDelete();
     * ```
     */
    const handleConfirmDelete = useCallback(() => {
        if (userToDelete) {
            setUsers(prev => prev.filter(u => u.id !== userToDelete.id));
            setDeletedMsg('User deleted');
            setTimeout(() => setDeletedMsg(null), 1800);
        }
        setConfirmOpen(false);
        setUserToDelete(null);
    }, [userToDelete]);

    /**
     * Cancels the delete operation and closes confirmation modal.
     * 
     * @example
     * ```typescript
     * handleCancelDelete();
     * ```
     */
    const handleCancelDelete = useCallback(() => {
        setConfirmOpen(false);
        setUserToDelete(null);
    }, []);

    /**
     * Memoized table rows to prevent unnecessary re-renders.
     * 
     * Each row includes:
     * - User name and email
     * - Formatted address
     * - Clickable phone number
     * - Clickable website link
     * - Company name
     * - Delete button with accessibility features
     * 
     * @type {JSX.Element[]}
     */
    const tableRows = useMemo(() => {
        return users.map((user) => (
            <tr
                key={user.id}
                className="table-row"
                onClick={() => handleRowClick(user)}
            >
                <td className="user-info">
                    <div className="user-name">{user.name}</div>
                    <div className="user-email">{user.email}</div>
                </td>
                <td className="user-address" title={formatAddress(user.address)}>
                    {formatAddress(user.address)}
                </td>
                <td className="user-phone">
                    <a
                        href={`tel:${user.phone}`}
                        onClick={(e) => e.stopPropagation()}
                    >
                        {user.phone}
                    </a>
                </td>
                <td className="user-website">
                    <a
                        href={formatWebsite(user.website)}
                        target="_blank"
                        rel="noopener noreferrer"
                        onClick={(e) => e.stopPropagation()}
                    >
                        {user.website}
                    </a>
                </td>
                <td className="user-company">
                    {user.company.name}
                </td>
                <td className="user-action">
                    <button
                        className="action-btn"
                        aria-label={`Remove user ${user.name}`}
                        tabIndex={0}
                        onClick={(e) => handleDeleteUser(user, e)}
                    >
                        <span className="action-x">Ã—</span>
                    </button>
                </td>
            </tr>
        ));
    }, [users, handleRowClick, handleDeleteUser]);

    return (
        <>
            <div className="user-list-container">
                <h2 className="user-list-title">User Directory</h2>
                <div className="user-table-wrapper">
                    <table className="user-table modern">
                        <thead>
                            <tr className="table-header">
                                <th>NAME / EMAIL</th>
                                <th>ADDRESS</th>
                                <th>PHONE</th>
                                <th>WEBSITE</th>
                                <th>COMPANY</th>
                                <th>ACTION</th>
                            </tr>
                        </thead>
                        <tbody>
                            {tableRows}
                        </tbody>
                    </table>
                </div>
                {deletedMsg && (
                    <div className="user-toast" role="status">{deletedMsg}</div>
                )}
                <ConfirmDeleteModal
                    open={confirmOpen}
                    onCancel={handleCancelDelete}
                    onConfirm={handleConfirmDelete}
                    userName={userToDelete?.name || ''}
                />
            </div>

            {isModalOpen && (
                <Suspense fallback={
                    <div className="modal-backdrop">
                        <div className="modal-container">
                            <div className="loading-container">
                                <div className="loading-spinner"></div>
                                <p>Loading modal...</p>
                            </div>
                        </div>
                    </div>
                }>
                    <UserDetailsModal
                        user={selectedUser}
                        isOpen={isModalOpen}
                        onClose={handleCloseModal}
                    />
                </Suspense>
            )}
        </>
    );
});

UserList.displayName = 'UserList';

export default UserList; 