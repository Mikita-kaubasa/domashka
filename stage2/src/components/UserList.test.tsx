import { fireEvent, render, screen } from '@testing-library/react';
import { User } from '../types/types';
import UserList from './UserList';

const mockUser: User = {
    id: 1,
    name: 'Alice',
    username: 'alice123',
    email: 'alice@example.com',
    address: {
        street: 'Main St',
        suite: 'Apt 1',
        city: 'Wonderland',
        zipcode: '12345',
        geo: { lat: '0', lng: '0' },
    },
    phone: '123-456-7890',
    website: 'alice.com',
    company: {
        name: 'Alice Co',
        catchPhrase: 'Curious and Curiouser',
        bs: 'adventures',
    },
};

const mockUsers: User[] = [
    mockUser,
    {
        ...mockUser,
        id: 2,
        name: 'Bob',
        username: 'bob321',
        email: 'bob@example.com',
        company: { ...mockUser.company, name: 'Bob Inc' },
    },
];

describe('UserList', () => {
    it('renders without crashing', () => {
        render(<UserList users={mockUsers} />);
        expect(screen.getByText('Alice')).toBeInTheDocument();
        expect(screen.getByText('Bob')).toBeInTheDocument();
    });

    it('renders correct number of users', () => {
        render(<UserList users={mockUsers} />);
        // Select all rows except the header
        const rows = screen.getAllByRole('row');
        // First row is header, rest are users
        expect(rows.length - 1).toBe(mockUsers.length);
    });

    it('renders empty table if no users', () => {
        render(<UserList users={[]} />);
        // There should be no table rows except the header
        const rows = screen.getAllByRole('row');
        // Only header row should exist
        expect(rows.length).toBe(1);
    });

    it('opens modal when a user row is clicked', () => {
        render(<UserList users={mockUsers} />);
        const aliceRow = screen.getByText('Alice').closest('tr');
        if (aliceRow) fireEvent.click(aliceRow);
        // Modal fallback or modal content should appear
        expect(screen.getByText(/loading modal/i)).toBeInTheDocument();
    });

    it('has correct accessibility attributes on delete buttons', () => {
        render(<UserList users={mockUsers} />);
        const deleteButtons = screen.getAllByRole('button', { name: /remove user/i });
        deleteButtons.forEach((btn) => {
            expect(btn).toHaveAttribute('tabindex', '0');
            expect(btn).toHaveAttribute('aria-label');
        });
    });

    it('matches snapshot', () => {
        const { asFragment } = render(<UserList users={mockUsers} />);
        expect(asFragment()).toMatchSnapshot();
    });
}); 