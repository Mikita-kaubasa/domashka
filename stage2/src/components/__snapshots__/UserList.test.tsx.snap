// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`UserList matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="user-list-container"
  >
    <h2
      class="user-list-title"
    >
      User Directory
    </h2>
    <div
      class="user-table-wrapper"
    >
      <table
        class="user-table modern"
      >
        <thead>
          <tr
            class="table-header"
          >
            <th>
              NAME / EMAIL
            </th>
            <th>
              ADDRESS
            </th>
            <th>
              PHONE
            </th>
            <th>
              WEBSITE
            </th>
            <th>
              COMPANY
            </th>
            <th>
              ACTION
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="table-row"
          >
            <td
              class="user-info"
            >
              <div
                class="user-name"
              >
                Alice
              </div>
              <div
                class="user-email"
              >
                alice@example.com
              </div>
            </td>
            <td
              class="user-address"
              title="Main St, Apt 1, Wonderland, 12345"
            >
              Main St, Apt 1, Wonderland, 12345
            </td>
            <td
              class="user-phone"
            >
              <a
                href="tel:123-456-7890"
              >
                123-456-7890
              </a>
            </td>
            <td
              class="user-website"
            >
              <a
                href="https://alice.com"
                rel="noopener noreferrer"
                target="_blank"
              >
                alice.com
              </a>
            </td>
            <td
              class="user-company"
            >
              Alice Co
            </td>
            <td
              class="user-action"
            >
              <button
                aria-label="Remove user Alice"
                class="action-btn"
                tabindex="0"
              >
                <span
                  class="action-x"
                >
                  ×
                </span>
              </button>
            </td>
          </tr>
          <tr
            class="table-row"
          >
            <td
              class="user-info"
            >
              <div
                class="user-name"
              >
                Bob
              </div>
              <div
                class="user-email"
              >
                bob@example.com
              </div>
            </td>
            <td
              class="user-address"
              title="Main St, Apt 1, Wonderland, 12345"
            >
              Main St, Apt 1, Wonderland, 12345
            </td>
            <td
              class="user-phone"
            >
              <a
                href="tel:123-456-7890"
              >
                123-456-7890
              </a>
            </td>
            <td
              class="user-website"
            >
              <a
                href="https://alice.com"
                rel="noopener noreferrer"
                target="_blank"
              >
                alice.com
              </a>
            </td>
            <td
              class="user-company"
            >
              Bob Inc
            </td>
            <td
              class="user-action"
            >
              <button
                aria-label="Remove user Bob"
                class="action-btn"
                tabindex="0"
              >
                <span
                  class="action-x"
                >
                  ×
                </span>
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
